Adding switch to submit on enter behaviour to <SimpleForm> and <TabbedForm>

jkrnak opened the pull request:
I needed this functionality because I've added `react-google-maps` with the `<Searchbox/>` to my form. But if the user was selecting from the searchbox by pressing enter on the keyboard and not by clicking on one of the results, the form got submitted. 

I found this comment https://github.com/erikras/redux-form/issues/572#issuecomment-268905929 on how to disable this auto submit behaviour.

~~I added `simple-mock` as a dependency as I needed spying on methods, I chose this one over `sinon` because this is smaller and has no dependencies. If you have anything against this library or if I have missed that there is already a mocking lib included I can amend the code.~~
I realised `sinon` was already added as a dependency.

This change doesn't break backward compatibility.


ps: awesome project! great work guys!

fzaninotto commented on src/mui/button/SaveButton.js:
```diff
@@ -49,6 +52,12 @@ SaveButton.propTypes = {
+};
+
+SaveButton.defaultProps = {
+    submitOnEnter: true,
```
you repeat the default value in too many places. It should only be set once.

jkrnak replied:
I feel it is something that belongs more to the form than the button, but even than I will need to set it on <SimpleForm> and <TabbedForm> too. Is that okay with you?

fzaninotto replied:
I'm not sure I understand you. But I agree that if there is a place to put the default, it's the form.

jkrnak replied:
I have no idea what I wanted to say :) That's not an English sentence.

fzaninotto commented on src/mui/button/SaveButton.js:
```diff
@@ -6,20 +6,23 @@ import ContentSave from 'material-ui/svg-icons/content/save';
             // prevent double submission
             e.preventDefault();
+        } else if (!this.props.submitOnEnter && this.props.handleSubmit) {
+            this.props.handleSubmit();
```
why not use redux-form's `submit()` action creator instead? That way, you don't have to pass `handleSubmit()` all the way down.

jkrnak replied:
Great idea! To do this I need to know the name of the form, which I'm not sure how to get. I've seen `record-form` and `record-form-<index>` namings as well. 
Now a see that probably the indexed naming is the new one although I don't see naming on the create forms, only on the edit ones. 
- Should there be explicit names defined for the create forms?
- Should I pass down the form name to this component as I don't see how I could get it in any other way?

fzaninotto replied:
`record-form-[index]` doesn't exist anymore (see #533), so you can safely use `record-form` everywhere (valid for both Create and Edit forms).

fzaninotto commented on src/mui/form/SimpleForm.js:
```diff
@@ -6,18 +6,22 @@ import getDefaultValues from './getDefaultValues';
-);
+export const SimpleForm = ({ children, handleSubmit, invalid, record, resource, basePath, submitOnEnter }) => {
+    const formOnSubmit = submitOnEnter ? handleSubmit : function noop() {};
+    const toolbarProps = submitOnEnter ? {} : { submitOnEnter: submitOnEnter, handleSubmit: handleSubmit };
```
I'd prefer you pass these directly in `<Toolbar>`, in JSX.

fzaninotto commented on src/mui/form/SimpleForm.js:
```diff
@@ -6,18 +6,22 @@ import getDefaultValues from './getDefaultValues';
-    </form>
-);
+export const SimpleForm = ({ children, handleSubmit, invalid, record, resource, basePath, submitOnEnter }) => {
+    const formOnSubmit = submitOnEnter ? handleSubmit : function noop() {};
```
this will create a new `noop()` function for every render, and therefore force a rerender of children even if nothing changed. You should move `noop()` definition out of the component.

Also, no need for a `const`, you can do that directly in the `onSubmit` prop, in JSX.

jkrnak commented:
@fzaninotto I've made the amendments you've asked for and rebased the branch and fixed all the conflicts. Is there anything else do you want me to do on this PR?

djhi commented on src/mui/button/SaveButton.js:
```diff
@@ -1,26 +1,30 @@
             // prevent double submission
             e.preventDefault();
+        } else if (!this.props.submitOnEnter) {
+            this.props.dispatch(submit('record-form'));
```
You should inject a `submit` function inside `connect` `mapDispatchToProps`.

```js
const mapDispatchToProps = ({ submit: () => submit('record-form') });

export default connect(mapStateToProps, mapDispatchToProps)(translate(SaveButton));
```

Then you'll just use `this.props.submit()`

djhi replied:
It's a good thing to avoid passing dispatch directly in the components. It makes them ignorant about how the submit itself is done

jkrnak replied:
Thanks for the feedback! It is much cleaner now.

djhi commented on src/mui/button/SaveButton.js:
```diff
@@ -1,26 +1,30 @@
         if (this.props.saving) {
             // prevent double submission
             e.preventDefault();
+        } else if (!this.props.submitOnEnter) {
```
If `submitOnEnter` is `true`, we can't submit by clicking the button ?

jkrnak replied:
if `submitOnEnter === true` the button type is `submit`, see line 24. So it falls back to how it used to work. We don't need to explicitly dispatch a `submit` action.

Should I add a comment there? This might not be obvious at first sight. I had to think about it too.

djhi replied:
A comment would be nice indeed :)

jkrnak commented:
One of the e2e test is flaky: https://travis-ci.org/marmelab/admin-on-rest/builds/233107073?utm_source=github_status&utm_medium=notification
Exact same commit broke it, I've only amended the message so I can force push and trigger a new one.

fzaninotto commented:
Excellent, thanks a lot!
